### ISSUE TITLE
Type Narrowing: `if val == Literal` should always narrow to Literal

### ISSUE THREAD

**Issue by CarliJoy** [python/mypy#16819]
**Feature**

Currently mypy already type narrows literals in if

```python
def works(abc: Literal["a", "b", "c"]):
    a: Literal["a"] | None = None
    if abc == "a":
        a = abc

def works_not(anystring: str):
    a: Literal["a"] | None = None
    if anystring == "a":
        a = anystring  # expression has type str
```

[More Complete MyPy Play](https://mypy-play.net/?mypy=master&python=3.12&gist=51c2c1423471c3f8e574705da5c52317)

**Pitch**

See #16813 for an example why this is useful.
In general working with TypedDicts  in for loop this helps a lot reducing duplicate code / keys.

**Comment by JelleZijlstra** [comment]
Agree. There's some risk this could lead to new false positives in cases involving invariance, but on balance this is almost certainly a positive.

**Comment by erictraut** [comment]
FWIW, pyright applies narrowing in this case. [Pyright playground](https://pyright-play.net/?pythonVersion=3.12&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2AsAFB10AmJwUA7rgNYDOAFOQCMAxgC5CxMlQDaAInIyANFBkDFyoTIC6AShF0oBzK0FCoAXjPL5e2obtQQJAG4kqAfXgIS-Ydoa1mVg4QHjcUMBh%2BFDhuGBBUNDFYkF19QyRjaOSE80s5GRt7Q0cXd09vciy4hL96WiA). I haven't observed false positives in cases involving invariance, but I agree that could be a small concern if code bases are relying on mypy's current behavior.

**Comment by CarliJoy** [comment]
Actually the full example works already in [pyright](https://pyright-play.net/?pythonVersion=3.12&code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQDRQAySMhIljAkmxwEYVCAWABQogCaFgUMCgpwqZPgApilAFzNW7SgG0ARGX2N9ffQF0AlOtFQ7UAHRPRLkZOkRiCAIzKAzjAgqGh%2Bmjw6AoS6ASBWNiL2UMC4mJgoUDHBobaJ9kjSSGlQyobGUKb61jm5uUSklDSIhMpIlnYAxBmB1TX2svKKKq1QnVCAMuRdIFAoYPhavHqlJmbmPYn5qQC8m%2BVG8b25-QpKLW2jEzHTs3YsCxQGRssWru5QnggATP6BWWELkdFAnEemoKKFYABXBCCXS3HT3UrmRhwjgIlbmKA7EqPcpmSw9ZJTQqoSa-NZ5ApFUHZBIHex1cjUWjNYYjOwXQJXOYoxY4iqrWl0mRyY5DfGC3IbQrbXb6fZC4UDE6s86TLk3bSoh5lfnkuwSKRvLwAZm%2BQXQ4PC-BhMWBEupmgoSACsM1vJ16MxUG1TwFiUJqRJmQt8pqUqplBpCoIJEZjTopzZMT1hxFg0TqsuM25boRfJWKfDMtKobpR3TKvZauzGruPtxzwlBxTiQNHi8ABYzb8oFalDCeXmPTqAMYWO3%2BlLE9LBkKl9aUknY4dVJu9BkNZkZyYtvpp5VtNnjavXebw%2Bu6td2Is7Eu7uzlg8dKAAd1wAGs-C9De8AKzdkNe3%2BAdcwvMwTDHcwJ3sB1IWhKJB21JEzy1fkMSxJYG3FSciSKWco16cMSQde8Y3qJkmm3ZMr1TJUxWfY8s1PRDMMvA4b1ledekfei2TfEBPxTNsjQQAA2AC5z7AEWPzUdx1XRJYKdF0ZOHKCvTAxscMDGcfhDQtF3SEiaMSDcKITYZOmo6MeMTI8OSmGsUPdX0DK2W89lI2zK34z9WxM%2BxSOE3y-GUJQR00VSnggiwFPsMhItAxEoAAHygAA5WRCC9TKUGECUNnCzEPLlFMyC9cLv2kEKqGzMKUDgWdNBiOK7AS5yhwxNLcuynYep6QqGtnYrOLKiqhr0jA2UIAAPBAiD8PwkFkKAAAsyD8WAmh3EQgA) as expected
